goth_tests/e2e/vm/assets/test_e2e_outbound_perf/image/outbound-bench/src/main.rs

```diff
@@ -0,0 +1,232 @@
+    });
+
+    let timeout_s = (mib as f32) / mib_per_s;
+    thread::sleep(Duration::from_secs_f32(timeout_s));
+
+    if !read_handle.is_finished() {
+        return Ok(false);
+    }
+
+    let recv_data = read_handle.join().unwrap()?;
+
+    send_handle.join().unwrap()?;
+
+    Ok(recv_data == data)
+}
+
+fn test_stress(mib_per_s: f32, stream: &impl Fn() -> TcpStream) -> anyhow::Result<bool> {
+    let mut data = vec![0u8; (mib_per_s * 1024.0 * 1024.0) as usize * 8];
+    rand::thread_rng().fill_bytes(&mut data);
+
+    let mut stream = stream();
+
+    let tries = 4;
+    for _ in 0..tries {
+        stream.write_all(&data)?;
+        thread::sleep(Duration::from_secs(10));
+    }
+
+    Ok(true)
+}
+
+fn test_iperf3(mib_per_s: f32, host: &str, port: u16) -> anyhow::Result<bool> {
+    let mut iperf3 = process::Command::new("iperf3")
+        .arg("-p")
+        .arg(port.to_string())
+        .arg("--json")
+        .arg("-c")
+        .arg(host)
+        .arg("-t")
+        .arg("5")
+        .arg("--logfile")
+        .arg("iperf.log")
+        .spawn()?;
+
+    iperf3.wait()?;
+
+    let log_text = std::fs::read_to_string("iperf.log")?;
+
+    let mut first_json = String::new();
+    let mut lines = log_text.lines();
+    first_json.push_str(lines.next().unwrap());
+    for line in lines {
+        if line.starts_with('{') {
+            break;
+        } else {
+            first_json.push_str(line);
+        }
+    }
+
+    let json: Value = serde_json::from_str(&first_json)?;
+
+    let bits_per_second = move || -> Option<f64> {
+        json.as_object()?
+            .get("end")?
+            .as_object()?
+            .get("sum_sent")?
+            .as_object()?
+            .get("bits_per_second")?
+            .as_f64()
+    }()
+    .ok_or(anyhow::anyhow!("malformed json"))?;
+
+    Ok(bits_per_second as f32 >= mib_per_s * 1024.0 * 1024.0 * 8.0)
+}
+
+fn test_many_reqs(total_reqs: usize, max_secs: f32) -> anyhow::Result<bool> {
+    let requests = [
+        "https://s3.amazonaws.com/data-production-walltime-info/production/dynamic/walltime-info.json?now=1528962473468.679.0000000000873",
+        "http://worldtimeapi.org/api/timezone",
+        "https://timeapi.io/api/Time/current/zone?timeZone=Europe/Amsterdam",
+        "http://www.7timer.info/bin/api.pl?lon=113.17&lat=23.09&product=astro&output=xml",
+    ];
+
+    let mut requests_to_run = Vec::new();
+    for request in requests {
+        for _ in 0..(total_reqs / requests.len()) {
+            requests_to_run.push(request);
+        }
+    }
+    while requests_to_run.len() < total_reqs {
+        requests_to_run.push(requests[0]);
+    }
+
+    requests_to_run.shuffle(&mut rand::thread_rng());
+
+    let started_at = Instant::now();
+    let result = tokio::runtime::Runtime::new()
+        .unwrap()
+        .block_on(async move {
+            let mut set = JoinSet::new();
+            for url in requests_to_run {
+                set.spawn(reqwest::get(url));
```

Comment:
you can use "blocking" feature in reqwest crate so you probably don't have to use tokio runtime here

Reply:
I opted to use `tokio` to run all of those requests asynchronously instead of spawning a thread for each of them. Notice that they run at the same time using the `JoinSet`.